### PR TITLE
server: restructure with options/config/server pattern

### PR DESCRIPTION
**Pull request by sttts**
Split up `Server.Run` into the usual `Config`, `completeConfig`, `Server`, `Server.Run` steps for modularity and easier customization and use in different binaries.

### CODE REVIEW

File: pkg/admission/initializers/interfaces.go [hunk 1]
```diff
@@ -32,13 +32,13 @@ type WantsKcpInformers interface {
...
 type WantsKubeClusterClient interface {
-	SetKubeClusterClient(kubeClusterClient *kubernetes.Cluster)
+	SetKubeClusterClient(kubeClusterClient kubernetes.ClusterInterface)
 }
 
 // WantsKcpClusterClient interface should be implemented by admission plugins
 // that want to have a kcp cluster client injected.
 type WantsKcpClusterClient interface {
-	SetKcpClusterClient(kubeClusterClient *kcpclientset.Cluster)
+	SetKcpClusterClient(kubeClusterClient kcpclientset.ClusterInterface)
```
Comment (stevekuznetsov) [inline pkg/admission/initializers/interfaces.go:41]: This seems like an unrelated change we could merge first/separately?

File: pkg/cliplugins/workload/plugin/syncer.yaml [hunk 2]
```diff
@@ -13,6 +13,15 @@ metadata:
   name: {{.ServiceAccount}}
   namespace: {{.Namespace}}
 ---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{.ServiceAccount}}-token
+  namespace: {{.Namespace}}
+  annotations:
+    kubernetes.io/service-account.name: {{.ServiceAccount}}
```
Comment (stevekuznetsov) [inline pkg/cliplugins/workload/plugin/syncer.yaml:22]: I still do not understand this whatsoever - this is making a token on the pcluster, right? We specifically don't want it to be an SA token, or annotated, right? Since if it is, the downstream controller deletes it?

File: pkg/embeddedetcd/config.go [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2021 The KCP Authors.
+Copyright 2022 The KCP Authors.
```
Comment (stevekuznetsov) [inline pkg/embeddedetcd/config.go:2]: Do we have to do this?

File: pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_controller.go [hunk 4]
```diff
@@ -53,20 +53,20 @@ const (
 func NewController(
 	kcpClusterClient kcpclient.Interface,
 	workspaceInformer tenancyinformer.ClusterWorkspaceInformer,
-	rootWorkspaceShardInformer tenancyinformer.ClusterWorkspaceShardInformer,
+	clusterWorkspaceShardInformer tenancyinformer.ClusterWorkspaceShardInformer,
```
Comment (stevekuznetsov) [inline pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_controller.go:56]: This seems like an unrelated change?

File: pkg/server/controllers.go [hunk 5]
```diff
@@ -214,19 +208,17 @@ func (s *Server) installKubeServiceAccountTokenController(ctx context.Context, c
 		return fmt.Errorf("error creating service account controller: %w", err)
 	}
 
-	s.AddPostStartHook("kcp-start-kube-service-account-token-controller", func(hookContext genericapiserver.PostStartHookContext) error {
+	return s.AddPostStartHook("kcp-start-kube-service-account-token-controller", func(hookContext genericapiserver.PostStartHookContext) error {
```
Comment (stevekuznetsov) [inline pkg/server/controllers.go:211]: This return pattern seems like another independent change?

File: pkg/server/options/options.go [hunk 6]
```diff
@@ -65,7 +66,7 @@ type ExtraOptions struct {
 
 type completedOptions struct {
 	GenericControlPlane options.CompletedServerRunOptions
-	EmbeddedEtcd        EmbeddedEtcd
+	EmbeddedEtcd        etcdoptions.CompletedObjects
```
Comment (stevekuznetsov) [inline pkg/server/options/options.go:69]: CompletedOptions?

File: pkg/server/server.go [hunk 7]
```diff
@@ -69,193 +71,194 @@ const resyncPeriod = 10 * time.Hour
...
+	kcpAdminToken, shardAdminToken string // nolint:structcheck
+	shardAdminTokenHash            []byte // nolint:structcheck
+
+	// clients
+	DynamicClusterClient       dynamic.ClusterInterface
+	KubeClusterClient          kubernetes.ClusterInterface
+	ApiExtensionsClusterClient apiextensionsclient.ClusterInterface
+	KcpClusterClient           kcpclient.ClusterInterface
+
+	// stuff
```
Comment (stevekuznetsov) [inline pkg/server/server.go:103]: ?

File: pkg/server/server.go [hunk 8]
```diff
@@ -69,193 +71,194 @@ const resyncPeriod = 10 * time.Hour
...
-	rootKcpSharedInformerFactory  kcpexternalversions.SharedInformerFactory
-	rootKubeSharedInformerFactory coreexternalversions.SharedInformerFactory
+type ExtraConfig struct {
+	// resolveIdenties is to be called on server start until it succeeds. It injects the kcp
+	// resource identities into the rest.Config used by the client. Only after it succeeds,
+	// the clients can wildcard-list/watch most kcp resources.
+	resolveIdentities func(ctx context.Context) error // nolint:structcheck
+
+	// authentication
+	kcpAdminToken, shardAdminToken string // nolint:structcheck
```
Comment (stevekuznetsov) [inline pkg/server/server.go:94]: IMO just turn off the check at the linter level, so when they fix their bugs and we turn it back on we get the full checking?

Comment (stevekuznetsov) [reply pkg/server/server.go:94]: Not sure we want to chase their false positives across the codebase (is it even deterministic?)

File: pkg/server/server.go [hunk 9]
```diff
@@ -498,12 +528,14 @@ func (s *Server) Run(ctx context.Context) error {
 		}
 
 		return nil
-	})
+	}); err != nil {
+		return err
+	}
 
 	// ========================================================================================================
 	// TODO: split apart everything after this line, into their own commands, optional launched in this process
```
Comment (stevekuznetsov) [inline pkg/server/server.go:536]: Can we remove this TODO unless we actually plan to do it?

File: pkg/server/server.go [hunk 10]
```diff
@@ -553,100 +585,66 @@ func (s *Server) Run(ctx context.Context) error {
 		}
 	}
 
-	if s.options.HomeWorkspaces.Enabled {
+	if s.Options.HomeWorkspaces.Enabled {
```
Comment (stevekuznetsov) [inline pkg/server/server.go:588]: not your change: but why does enableAll not trigger this one?

File: pkg/admission/initializers/interfaces.go (hunk 1 above)
Comment (sttts) [reply pkg/admission/initializers/interfaces.go:41]: no, it was necessary because we store the ClusterInterface in the config, not the client pointer.

File: pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_controller.go (hunk 4 above)
Comment (sttts) [reply pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_controller.go:56]: no, we had `rootWorkspaceShardInformer` in the server for reasons that must have disappeared. I removed the `root*` variant of those informers. Hence, also here.

File: pkg/server/controllers.go (hunk 5 above)
Comment (sttts) [reply pkg/server/controllers.go:211]: no. I unified the `Server.AddPostStartHook` signature and lint complained here.

File: pkg/server/options/options.go (hunk 6 above)
Comment (sttts) [reply pkg/server/options/options.go:69]: :D

Comment (sttts) [reply pkg/server/options/options.go:69]: fixed

File: pkg/server/server.go (hunk 8 above)
Comment (sttts) [reply pkg/server/server.go:94]: If I just bump the version?

File: pkg/server/server.go (hunk 7 above)
Comment (sttts) [reply pkg/server/server.go:103]: misc

File: pkg/server/server.go (hunk 8 above)
Comment (sttts) [reply pkg/server/server.go:94]: done

File: pkg/server/server.go (hunk 9 above)
Comment (sttts) [reply pkg/server/server.go:536]: oh, this will be a follow-up. Part of this PR is to make that possible.

File: pkg/server/server.go (hunk 10 above)
Comment (sttts) [reply pkg/server/server.go:588]: it's not a controller, but a feature. I would actually propose to remove `EnableAll` and only allow opt-out (in follow-up).

File: pkg/embeddedetcd/config.go (hunk 3 above)
Comment (sttts) [reply pkg/embeddedetcd/config.go:2]: it's technically a new file, only Github shows it as rename.

File: pkg/embeddedetcd/server.go [hunk 11]
```diff
@@ -31,6 +31,9 @@ type Server struct {
 }
 
 func NewServer(config CompletedConfig) *Server {
+	if config.Config == nil {
+		return nil
```
Comment (p0lyn0mial) [inline pkg/embeddedetcd/server.go:35]: shouldn't we handle a nil pointer in kcp.go:L134 then?

File: pkg/server/controllers.go [hunk 12]
```diff
@@ -374,19 +362,19 @@ func (s *Server) installWorkspaceScheduler(ctx context.Context, config *rest.Con
 
 	workspaceController, err := clusterworkspace.NewController(
 		kcpClusterClient,
-		s.kcpSharedInformerFactory.Tenancy().V1alpha1().ClusterWorkspaces(),
-		s.rootKcpSharedInformerFactory.Tenancy().V1alpha1().ClusterWorkspaceShards(),
-		s.kcpSharedInformerFactory.Apis().V1alpha1().APIBindings(),
+		s.KcpSharedInformerFactory.Tenancy().V1alpha1().ClusterWorkspaces(),
+		s.KcpSharedInformerFactory.Tenancy().V1alpha1().ClusterWorkspaceShards(),
```
Comment (p0lyn0mial) [inline pkg/server/controllers.go:366]: am I right that `ClusterWorkspaceShards` are only in the root `ws`?

File: pkg/reconciler/tenancy/clusterworkspaceshard/clusterworkspaceshard_controller.go [hunk 13]
```diff
@@ -47,18 +47,18 @@ const (
...
 ) (*Controller, error) {
 	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-workspaceshard")
 
 	c := &Controller{
-		queue:                     queue,
-		kcpClient:                 rootKcpClient,
-		rootWorkspaceShardIndexer: rootWorkspaceShardInformer.Informer().GetIndexer(),
-		rootWorkspaceShardLister:  rootWorkspaceShardInformer.Lister(),
+		queue:                        queue,
+		kcpClient:                    rootKcpClient,
```
Comment (p0lyn0mial) [inline pkg/reconciler/tenancy/clusterworkspaceshard/clusterworkspaceshard_controller.go:56]: so, this is now `kcpClusterClient`

File: pkg/server/controllers.go (hunk 12 above)
Comment (sttts) [reply pkg/server/controllers.go:366]: yes. But for that we should use informer scoping if we want that.

File: pkg/embeddedetcd/server.go (hunk 11 above)
Comment (sttts) [reply pkg/embeddedetcd/server.go:35]: 134 is about the kcp server, not etcd server.
